tokio/src/sync/notify.rs

```diff
@@ -319,42 +347,108 @@ impl Notify {
     }
 
     /// Notifies all waiting tasks
-    pub(crate) fn notify_waiters(&self) {
+    ///
+    /// If a task is currently waiting, that task is notified. Unlike with
+    /// `notify()`, no permit is stored to be used by the next call to
+    /// [`notified().await`]. The purpose of this method is to notify all
+    /// already registered waiters. Registering for notification is done by
+    /// acquiring an instance of the `Notified` future via calling `notified()`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     let notified1 = notify.notified();
+    ///     let notified2 = notify.notified();
+    ///
+    ///     let handle = tokio::spawn(async move {
+    ///         println!("sending notifications");
+    ///         notify2.notify_waiters();
+    ///     });
+    ///
+    ///     notified1.await;
+    ///     notified2.await;
+    ///     println!("received notifications");
+    /// }
+    /// ```
+    pub fn notify_waiters(&self) {
+        const NUM_WAKERS: usize = 32;
+
+        let mut wakers: [Option<Waker>; NUM_WAKERS] = Default::default();
+        let mut curr_waker = 0;
```

Comment:
nit: i might move this until after the `if let` on line 394, since we might return early there. we could skip initializing the stack array in that case, since it won't be used. this probably doesn't have a *huge* performance impact but it might be worth doing.

Reply:
I would avoid complicating code for micro-optimizations unless it is measurable.

Reply:
i thought it also reads nicer if these variables are declared closer to where they're actually used, but :shrug: